File: erp/erp/fleet/page/monitoring_board/monitoring_board.py
import frappe
from frappe.utils import now_datetime, time_diff_in_seconds

from erp.fleet.vehicle_status import job_duration, status_colors, status_icons, status_map


@frappe.whitelist()
def get_rows():
    """1 baris per kendaraan aktif untuk tabel Monitoring.

    Job aktif memakai aturan yang sama dengan halaman GPS Vehicle: item Dispatch Order
    assigned yang belum ditekan Lanjut Job / Menuju Garasi.
    Driver = driver job unit itu, atau yang check in di unit itu; hanya ditampilkan
    kalau dia sudah absen hari ini (kosong = belum absensi).
    Keterangan = input user di Driver Attendance (Izin / Sakit / catatan) untuk driver
    di baris itu -- sumber yang sama dengan halaman Driver Monitor.
    Durasi = sejak driver di-assign job sampai sekarang (job di sini selalu masih
    berjalan; durasi finalnya kelihatan di tab History driver).
    Note = Monitoring Notes TERAKHIR unit itu. Notifikasi masih dihitung di sini
    (checkpoint / GPS diam / Standby) sampai doctype notifikasi vehicle dibuat.
    """
    frappe.has_permission("Vehicle", "read", throw=True)

    jobs = {}
    for r in frappe.db.sql(
        """select i.vehicle, i.driver, i.dpo_no, i.customer, i.atd, i.ata, do.name dpo, (
                 select min(t.start) from `tabDispatch Order Route` t
                 where t.dpo_item = i.name and t.step_type = 'Assign') assign, (
                 select t.point from `tabDispatch Order Route` t
                 where t.dpo_item = i.name and t.step_type = 'Route'
                   and (t.start is not null or t.end is not null)
                 order by t.step desc limit 1) checkpoint
           from `tabDispatch Order Item` i
           join `tabDispatch Order` do on i.parent = do.name
           where i.assigned = 1 and ifnull(i.vehicle, '') != ''
             and not exists (
               select 1 from `tabDispatch Order Route` t
               where t.dpo_item = i.name and t.step_type in ('Lanjut Job', 'Menuju Garasi')
                 and t.start is not null)
           order by do.creation desc""",
        as_dict=True,
    ):
        jobs.setdefault(r.vehicle, r)

    # notifikasi terakhir per unit dari feed Vehicle Notification
    notifs = {}
    for n in frappe.db.sql(
        """select vehicle, message, notification_date from `tabVehicle Notification`
           where ifnull(vehicle, '') != '' order by notification_date""",
        as_dict=True,
    ):
        notifs[n.vehicle] = n  # yang terakhir menang

    # Hanya driver yang absen hari ini yang boleh muncul (ada nama = sudah absensi).
    # Unit bisa beda antara absen dan job (absen di unit A, ditugaskan unit B), jadi
    # nama diambil dari driver job dulu, baru unit tempat dia check in.
    hadir = {}  # driver -> nama
    checkin = {}  # vehicle -> driver, check in terakhir hari ini yang menang
    ket = {}  # driver -> keterangan (Izin/Sakit + catatan) yang diinput user
    for a in frappe.db.sql(
        """select a.driver, a.type, a.vehicle, a.remark, ifnull(d.title, a.driver) nama
           from `tabDriver Attendance` a left join `tabDriver` d on d.name = a.driver
           where a.type in ('Absensi', 'Check In', 'Izin', 'Sakit')
             and date(a.timestamp) = curdate()
           order by a.timestamp""",
        as_dict=True,
    ):
        if a.type in ("Absensi", "Check In"):
            hadir[a.driver] = a.nama  # Izin/Sakit bukan bukti hadir, jangan masuk sini
        if a.type == "Check In" and a.vehicle:
            checkin[a.vehicle] = a.driver
        if a.type in ("Izin", "Sakit") or a.remark:
            ket[a.driver] = " - ".join(
                x for x in (a.type if a.type in ("Izin", "Sakit") else None, a.remark) if x
            )

    notes = {}
    for n in frappe.db.sql(
        """select vehicle, note, ifnull(note_date, creation) note_date
           from `tabMonitoring Notes` where ifnull(vehicle, '') != ''
           order by ifnull(note_date, creation)""",
        as_dict=True,
    ):
        notes[n.vehicle] = n  # yang terakhir menang

    now = now_datetime()
    statuses = status_map(jobs)
    rows = []
    for v in frappe.db.sql(
        """select v.name, v.title, v.branch, g.modified gps_time, g.latitude, g.longitude
           from `tabVehicle` v
           left join `tabGPS Vehicle` g on g.name = (
                select g2.name from `tabGPS Vehicle` g2 where g2.vehicle = v.name
                order by g2.modified desc limit 1)
           where ifnull(v.disabled, 0) = 0
           order by v.branch, v.title""",
        as_dict=True,
    ):
        job = jobs.get(v.name)
        if job:
            notif = f"Di {job.checkpoint}" if job.checkpoint else "Belum lapor titik"
        elif not v.gps_time:
            notif = "Belum ada GPS"
        else:
            mins = int(time_diff_in_seconds(now, v.gps_time) // 60)
            notif = f"GPS diam {_age(mins)}" if mins > 30 else "Standby"
        n = notifs.get(v.name)  # feed notifikasi menang atas hitungan di atas
        c = notes.get(v.name)
        drv = _driver(job, checkin.get(v.name), jobs)
        rows.append(
            {
                "vehicle": v.name,
                "branch": v.branch or "",
                "nopol": v.title,
                "driver": hadir.get(drv, ""),
                # id driver tetap dikirim walau namanya tidak tampil (belum absen):
                # justru driver itulah yang paling sering perlu ditandai Izin/Sakit.
                "driver_id": drv or "",
                "status": statuses.get(v.name, "Not Active"),
                "job_no": (job and job.dpo_no) or "",
                "dpo": (job and job.dpo) or "",
                "customer": (job and job.customer) or "",
                "atd": (job and job.atd) or "",
                "ata": (job and job.ata) or "",
                "durasi": job_duration(job.assign, now=now) if job else "",
                "keterangan": ket.get(drv, ""),
                "notifikasi": (n and n.message) or notif,
                "notification_date": (n and n.notification_date) or v.gps_time or "",
                "note": (c and c.note) or "",
                "note_date": (c and c.note_date) or "",
                "latitude": v.latitude,
                "longitude": v.longitude,
            }
        )
    # dict, bukan list: palet status ikut dikirim supaya warna badge cuma didefinisikan
    # sekali (di vehicle_status.py) untuk semua halaman.
    return {"rows": rows, "status_colors": status_colors(), "status_icons": status_icons()}


@frappe.whitelist()
def get_notifications(vehicle, dpo_no=None, limit=30):
    """Daftar notifikasi unit (feed Vehicle Notification) — dipanggil saat modal dibuka."""
    frappe.has_permission("Vehicle Notification", "read", throw=True)
    return frappe.db.sql(
        """select name, message, notification_date, type, point, dpo_no
           from `tabVehicle Notification`
           where vehicle = %s or (%s != '' and dpo_no = %s)
           order by notification_date desc limit %s""",
        (vehicle, dpo_no or "", dpo_no or "", int(limit)),
        as_dict=True,
    )


def _driver(job, checkin_driver, jobs):
    """Driver job unit ini; kalau tidak ada job, driver yang check in di unit ini —
    kecuali dia sudah tampil di unit job-nya (absen di unit A, ditugaskan unit B)."""
    if job:
        return job.driver
    if checkin_driver and any(j.driver == checkin_driver for j in jobs.values()):
        return None
    return checkin_driver


def _age(mins):
    return f"{mins // 60}j" if mins >= 60 else f"{mins}m"

File: erp/erp/fleet/page/monitoring_board/monitoring_board.js
// Monitoring: satu tabel semua kendaraan aktif (job, notifikasi, note terakhir).
// Auto-refresh 60 detik.
frappe.pages['monitoring-board'].on_page_load = function (wrapper) {
	const page = frappe.ui.make_app_page({
		parent: wrapper,
		title: __('Monitoring'),
		single_column: true,
	});

	const $body = $(`
		<div class="mb">
			<style>
				.mb { padding: 8px 0; }
				.mb-scroll { overflow: auto; border: 1px solid var(--border-color); border-radius: 8px; }
				.mb table { width: 100%; margin: 0; font-size: 12px; border-collapse: separate; border-spacing: 0; }
				.mb th { position: sticky; top: 0; z-index: 1; background: var(--bg-light-gray, #f3f4f6);
					font-size: 11px; font-weight: 700; white-space: nowrap; padding: 6px 8px;
					border-bottom: 1px solid var(--border-color); }
				.mb td { padding: 3px 8px; white-space: nowrap; max-width: 260px; overflow: hidden;
					text-overflow: ellipsis; border-bottom: 1px solid var(--border-color); }
				.mb tbody tr:hover { background: var(--highlight-color, #eef2ff); }
				.mb .mb-pill { padding: 0 7px; border-radius: 9px; font-size: 11px; font-weight: 600; }
				.mb .btn-xs { padding: 1px 6px; font-size: 11px; }
				.mb-act { white-space: nowrap; }
				.mb-act .btn { padding: 2px 5px; line-height: 1; }
				.mb-act .icon { width: 14px; height: 14px; vertical-align: middle; }
				.mb-empty { padding: 14px; color: var(--text-muted); font-size: 13px; }
				.leaflet-tooltip.mb-plate { background: #000; color: #fff; border: none; box-shadow: none;
					font-size: 11px; font-weight: 700; padding: 1px 6px; border-radius: 3px; }
				.leaflet-tooltip.mb-plate:before { display: none; }
				/* modal peta dibuat selebar layar supaya petanya lega */
				.modal.mb-wide .modal-dialog { max-width: 96vw; width: 96vw; }
				.mb-dlg { display: flex; gap: 14px; align-items: stretch; }
				.mb-dlg-left { flex: 1 1 0; min-width: 420px; }
				.mb-dlg-col { border-left: 1px solid var(--border-color); padding-left: 12px;
					display: flex; flex-direction: column; min-width: 0; }
				.mb-dlg-route { flex: 0 0 175px; }
				.mb-dlg-notif { flex: 0 0 195px; }
				.mb-dlg-hist { flex: 0 0 205px; }
				.mb-tag { display: inline-block; font-size: 10px; font-weight: 700; padding: 0 6px; border-radius: 8px; }
				.mb-tag-job { background: #dbeafe; color: #1e40af; }
				.mb-tag-idle { background: var(--bg-light-gray, #f3f4f6); color: var(--text-muted); }
				.mb-done { color: #16a34a; font-weight: 600; }
				.mb-cur { background: #1d4ed8; color: #fff; font-weight: 700; padding: 1px 6px; border-radius: 9px; }
				.mb-dlg-info { display: grid; grid-template-columns: repeat(4, 1fr); gap: 6px 12px; margin: 10px 0 0; }
				.mb-dlg-info > div { min-width: 0; }
				.mb-dlg-info dt { font-size: 10px; font-weight: 700; color: #374151; text-transform: uppercase; letter-spacing: .03em; }
				.mb-dlg-info dd { margin: 0; font-size: 12.5px; white-space: nowrap; overflow: hidden; text-overflow: ellipsis; }
			</style>
			<div class="mb-scroll"><table class="table">
				<thead><tr>
					<th>No</th><th>Brc</th><th>Nopol</th><th>Status</th><th>Job No</th>
					<th>Customer</th><th>ATD</th><th>ATA</th><th>Notifikasi</th><th>Notification Date</th>
					<th>Note</th><th>Note Date</th>
				</tr></thead>
				<tbody></tbody>
			</table></div>
		</div>
	`).appendTo(page.body);

	page.set_secondary_action(__('Refresh'), load);

	const esc = frappe.utils.escape_html;
	// Palet datang dari server (erp/fleet/vehicle_status.py) supaya status baru tidak perlu
	// didaftarkan ulang di tiap halaman.
	let STATUS_STYLE = {};
	const dt = (v) => (v ? frappe.datetime.str_to_user(String(v).slice(0, 19)) : '-');
	const tgl = (v) => (v ? frappe.datetime.str_to_user(String(v).slice(0, 10)) : '-');
	let rows = [];

	// tabel mengisi sisa layar
	function fit() {
		const top = $body.find('.mb-scroll').offset().top - $(window).scrollTop();
		$body.find('.mb-scroll').css('max-height', Math.max(window.innerHeight - top - 40, 160) + 'px');
	}

	function paint() {
		const $tb = $body.find('tbody').empty();
		if (!rows.length) {
			$tb.append(`<tr><td colspan="12" class="mb-empty">${__('Belum ada kendaraan aktif.')}</td></tr>`);
			return;
		}
		rows.forEach((r, i) => {
			$tb.append(`<tr data-i="${i}">
				<td>${i + 1}</td>
				<td>${esc(r.branch || '-')}</td>
				<td><a href="#" class="mb-open"><b>${esc(r.nopol)}</b></a></td>
				<td><span class="mb-pill" style="${STATUS_STYLE[r.status] || ''}">${esc(r.status)}</span></td>
				<td>${esc(r.job_no || '-')}</td>
				<td title="${esc(r.customer || '')}">${esc(r.customer || '-')}</td>
				<td>${tgl(r.atd)}</td>
				<td>${tgl(r.ata)}</td>
				<td title="${esc(r.notifikasi || '')}">${esc(r.notifikasi || '-')}</td>
				<td>${dt(r.notification_date)}</td>
				<td title="${esc(r.note || '')}">${esc(r.note || '-')}</td>
				<td>${dt(r.note_date)}</td>
			</tr>`);
		});
		$tb.find('.mb-open').on('click', function (e) {
			e.preventDefault();
			map_dialog(rows[$(this).closest('tr').data('i')]);
		});
		fit();
	}

	// "05 Agu 26 18:00 - 23:59" / "... - On Going" / "-"
	const BULAN = ['Jan', 'Feb', 'Mar', 'Apr', 'Mei', 'Jun', 'Jul', 'Agu', 'Sep', 'Okt', 'Nov', 'Des'];
	const tg = (s) => `${s.slice(8, 10)} ${BULAN[Number(s.slice(5, 7)) - 1]} ${s.slice(2, 4)}`;
	const jm = (s) => s.slice(11, 16);
	function in_out(a, b) {
		if (!a && !b) return '-';
		if (a && !b) return `${tg(a)} ${jm(a)} - ${__('On Going')}`;
		if (!a && b) return `? - ${tg(b)} ${jm(b)}`;
		return a.slice(0, 10) === b.slice(0, 10) ? `${tg(a)} ${jm(a)} - ${jm(b)}` : `${tg(a)} ${jm(a)} - ${tg(b)} ${jm(b)}`;
	}

	// titik rute job aktif unit ini (diambil saat modal dibuka)
	function load_route(r, $box) {
		const head = `<div style="font-weight:700;margin-bottom:6px">${__('Route Assigned')}</div>`;
		$box.html(`${head}<div class="text-muted" style="font-size:12px">${__('Memuat...')}</div>`);
		frappe
			.call('erp.fleet.page.gps_monitor.gps_monitor.get_detail', { vehicle: r.vehicle })
			.then((res) => {
				const pts = (res.message && res.message.route_points) || [];
				const cur = pts.findIndex((p) => !p.end);
				const isi = pts.length
					? pts
							.map(
								(p, i) => `<div style="padding:4px 0;border-bottom:1px solid var(--border-color)">
									<div style="font-size:12px" class="${p.end ? 'mb-done' : ''}">
										<b>${i + 1}.</b> ${i === cur ? `<span class="mb-cur">${esc(p.point)}</span>` : esc(p.point)}</div>
									<div style="font-size:10.5px;color:#374151;padding-left:15px">${in_out(p.start, p.end)}</div>
								</div>`
							)
							.join('')
					: `<div class="text-muted" style="font-size:12px">${
							r.dpo ? __('Belum ada titik rute.') : __('Tidak ada job aktif.')
					  }</div>`;
				$box.html(`${head}<div style="flex:1 1 auto;min-height:0;overflow-y:auto">${isi}</div>`);
			});
	}

	// feed notifikasi unit ini
	function load_notifications(r, $box) {
		const head = `<div style="font-weight:700;margin-bottom:6px">${__('Notifikasi')}</div>`;
		$box.html(`${head}<div class="text-muted" style="font-size:12px">${__('Memuat...')}</div>`);
		frappe
			.call('erp.fleet.page.monitoring_board.monitoring_board.get_notifications', {
				vehicle: r.vehicle,
				dpo_no: r.job_no || '',
			})
			.then((res) => {
				const list = res.message || [];
				const isi = list.length
					? list
							.map(
								(n) => `<div style="border-bottom:1px solid var(--border-color);padding:5px 0">
									<div style="font-size:12.5px">${esc(n.message)}</div>
									<div style="font-size:10.5px;color:var(--text-muted)">${dt(n.notification_date)}</div>
								</div>`
							)
							.join('')
					: `<div style="border-bottom:1px solid var(--border-color);padding:5px 0">
							<div style="font-size:12.5px">${esc(r.notifikasi || '-')}</div>
							<div style="font-size:10.5px;color:var(--text-muted)">${dt(r.notification_date)}</div>
						</div>`;
				$box.html(`${head}<div style="flex:1 1 auto;min-height:0;overflow-y:auto">${isi}</div>`);
			});
	}

	// penanda tiap note: dibuat saat unit punya job (tampilkan No DPO-nya) atau tidak
	const job_tag = (n) =>
		n.dpo_no
			? `<span class="mb-tag mb-tag-job" title="${__('Dibuat saat ada job')}">${esc(n.dpo_no)}</span>`
			: `<span class="mb-tag mb-tag-idle">${__('Tanpa job')}</span>`;

	// semua note unit itu (dengan atau tanpa job) — dipakai dialog peta & dialog note
	function load_history(r, $box) {
		const head = `<div style="font-weight:700;margin-bottom:6px">${__('Notes')}</div>`;
		$box.html(`${head}<div class="text-muted" style="font-size:12px">${__('Memuat...')}</div>`);
		frappe
			.call('erp.fleet.page.gps_monitor.gps_monitor.get_notes', {
				dpo_no: r.job_no || '',
				vehicle: r.vehicle,
			})
			.then((res) => {
				const list = res.message || [];
				const isi = list.length
					? list
							.map(
								(n) => `<div style="border-bottom:1px solid var(--border-color);padding:5px 0">
									<div style="font-size:11px;font-weight:700;color:#374151">${dt(n.note_date)} &middot; ${esc(n.owner)}</div>
									<div style="font-size:12.5px">${esc(n.note)}</div>
									<div>${job_tag(n)}</div>
								</div>`
							)
							.join('')
					: `<div class="text-muted" style="font-size:12px">${__('Belum ada note.')}</div>`;
				$box.html(`${head}<div style="flex:1 1 auto;min-height:0;overflow-y:auto">${isi}</div>`);
			});
	}

	// peta posisi 1 unit, plus tombol pindah ke halaman GPS Vehicle
	function map_dialog(r) {
		if (!r.latitude || !r.longitude) {
			frappe.msgprint(__('Unit {0} belum punya koordinat GPS.', [r.nopol]));
			return;
		}
		const d = new frappe.ui.Dialog({
			title: __('Posisi {0}', [r.nopol]),
			size: 'extra-large',
			fields: [{ fieldtype: 'HTML', fieldname: 'map' }],
			primary_action_label: __('Go to Map'),
			primary_action: () => {
				d.hide();
				frappe.route_options = { vehicle: r.vehicle };
				frappe.set_route('gps-monitor');
			},
			secondary_action_label: __('Add Note'),
			secondary_action: () => note_dialog(r),
		});
		d.$wrapper.addClass('mb-wide');
		d.show();
		setTimeout(() => {
			const $w = d.get_field('map').$wrapper;
			const pair = (label, val) =>
				`<div><dt>${label}</dt><dd title="${esc(val || '')}">${val || '-'}</dd></div>`;
			$w.html(`<div class="mb-dlg">
					<div class="mb-dlg-left">
						<div class="mb-dlg-map border rounded" style="height:560px"></div>
						<dl class="mb-dlg-info">
							${pair(__('Nopol'), `<b>${esc(r.nopol)}</b>`)}
							${pair(__('Branch'), esc(r.branch))}
							${pair(__('Status'), `<span class="mb-pill" style="${STATUS_STYLE[r.status] || ''}">${esc(r.status)}</span>`)}
							${pair(
								__('Job No'),
								r.dpo ? `<a href="#" class="mb-job">${esc(r.job_no)}</a>` : esc(r.job_no)
							)}
							${pair(__('Customer'), esc(r.customer))}
							${pair(__('ATD'), tgl(r.atd))}
							${pair(__('ATA'), tgl(r.ata))}
						</dl>
					</div>
					<div class="mb-dlg-col mb-dlg-route"></div>
					<div class="mb-dlg-col mb-dlg-notif"></div>
					<div class="mb-dlg-col mb-dlg-hist"></div>
				</div>`);
			$w.find('.mb-job').on('click', (ev) => {
				ev.preventDefault();
				d.hide();
				frappe.set_route('Form', 'Dispatch Order', r.dpo);
			});
			load_route(r, $w.find('.mb-dlg-route'));
			load_notifications(r, $w.find('.mb-dlg-notif'));
			load_history(r, $w.find('.mb-dlg-hist'));
			const m = L.map($w.find('.mb-dlg-map')[0]).setView([r.latitude, r.longitude], 14);
			L.tileLayer('https://{s}.basemaps.cartocdn.com/rastertiles/voyager/{z}/{x}/{y}{r}.png', {
				attribution: '&copy; OpenStreetMap, &copy; CARTO',
				subdomains: 'abcd',
				maxZoom: 20,
			}).addTo(m);
			L.marker([r.latitude, r.longitude], {
				icon: L.icon({
					iconUrl: '/assets/erp/images/truck.png',
					iconSize: [50, 50],
					iconAnchor: [25, 25],
					popupAnchor: [0, -20],
				}),
			})
				.addTo(m)
				.bindTooltip(esc(r.nopol), { permanent: true, direction: 'top', offset: [0, -22], className: 'mb-plate' });
			setTimeout(() => m.invalidateSize(), 300);
			d.$wrapper.on('hidden.bs.modal', () => m.remove());
		}, 300);
	}

	// note + history-nya, sama seperti di halaman GPS Vehicle
	function note_dialog(r) {
		const d = new frappe.ui.Dialog({
			title: __('Note {0}', [r.nopol]),
			size: 'large',
			fields: [
				{ fieldtype: 'HTML', options: `<div style="font-weight:700;margin-bottom:6px">${__('Note Anda')}</div>` },
				{ fieldname: 'note_date', fieldtype: 'Datetime', label: __('Note Date'), default: frappe.datetime.now_datetime(), reqd: 1 },
				{ fieldname: 'note', fieldtype: 'Small Text', label: __('Note Anda'), reqd: 1 },
				{ fieldtype: 'Column Break' },
				{ fieldtype: 'HTML', fieldname: 'history' },
			],
			primary_action_label: __('Submit'),
			secondary_action_label: __('Cancel'),
			secondary_action: () => d.hide(),
			primary_action: (v) => {
				frappe
					.call('erp.fleet.page.gps_monitor.gps_monitor.add_note', {
						vehicle: r.vehicle,
						nopol: r.nopol,
						dpo_no: r.job_no,
						status: r.status,
						note_date: v.note_date,
						note: v.note,
					})
					.then((res) => {
						d.hide();
						frappe.show_alert({ message: __('Note tersimpan: {0}', [res.message]), indicator: 'green' });
						load();
					});
			},
		});
		d.show();

		const $h = d.get_field('history').$wrapper;
		const head = `<div style="font-weight:700;margin-bottom:6px">${__('History Note')}</div>`;
		$h.html(`${head}<div class="text-muted" style="font-size:12px">${__('Memuat...')}</div>`);
		frappe
			.call('erp.fleet.page.gps_monitor.gps_monitor.get_notes', { dpo_no: r.job_no || '', vehicle: r.vehicle })
			.then((res) => {
				const list = res.message || [];
				const isi = list.length
					? list
							.map(
								(n) => `<div style="border-bottom:1px solid var(--border-color);padding:6px 0">
									<div style="font-size:11px;font-weight:700;color:#374151">${dt(n.note_date)}</div>
									<div style="font-size:12.5px">${esc(n.note)}</div>
									<div style="font-size:10.5px;color:var(--text-muted)">${esc(n.owner)}</div>
								</div>`
							)
							.join('')
					: `<div class="text-muted" style="font-size:12px">${__('Belum ada note.')}</div>`;
				$h.html(`${head}<div style="max-height:260px;overflow-y:auto">${isi}</div>`);
			});
	}

	function load() {
		frappe.call('erp.fleet.page.monitoring_board.monitoring_board.get_rows').then((r) => {
			const res = r.message || {};
			rows = res.rows || [];
			STATUS_STYLE = res.status_colors || STATUS_STYLE;
			paint();
		});
	}

	load();
	const timer = setInterval(load, 60000);
	$(window).on('resize.mb', fit);
	$(wrapper).on('remove', () => {
		clearInterval(timer);
		$(window).off('resize.mb');
	});
};
